url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/io/DiskStorageManager.java

```diff
@@ -0,0 +1,423 @@
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+//
+// Contact information:
+//  Mailing address:
+//    Marios Hadjieleftheriou
+//    University of California, Riverside
+//    Department of Computer Science
+//    Surge Building, Room 310
+//    Riverside, CA 92521
+//
+//  Email:
+//    marioh@cs.ucr.edu
+
+package spatialindex.io;
+
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeSet;
+
+import spatialindex.setting.PropertySet;
+
+public class DiskStorageManager implements IStorageManager {
+	private RandomAccessFile m_dataFile = null;
+	private RandomAccessFile m_indexFile = null;
+	private int m_pageSize = 0;
+	private int m_nextPage = -1;
+	/**
+	 * store the id of blank page
+	 */
+	private final TreeSet<Integer> m_emptyPages = new TreeSet<Integer>();
+	private final HashMap<Integer, Entry> m_pageIndex = new HashMap<Integer, Entry>();
+	/**
+	 * after init, it has the length of m_pageSize
+	 */
+	private byte[] m_buffer = null;
+
+	/**
+	 * ps should contain "Overwrite", "Filename", "Pagesize"
+	 * 
+	 * @param ps
+	 * @throws SecurityException
+	 * @throws NullPointerException
+	 * @throws IOException
+	 * @throws FileNotFoundException
+	 * @throws IllegalArgumentException
+	 */
+	public DiskStorageManager(PropertySet ps) throws SecurityException,
+			NullPointerException, IOException, FileNotFoundException,
+			IllegalArgumentException {
+		Object var;
+
+		// Open/Create flag.
+		boolean bOverwrite = false;
+		var = ps.getProperty("Overwrite");
+
+		if (var != null) {
+			if (!(var instanceof Boolean)) {
+				throw new IllegalArgumentException(
+						"Property Overwrite must be a Boolean");
+			}
+			bOverwrite = ((Boolean) var).booleanValue();
+		}
+
+		// storage filename.
+		var = ps.getProperty("FileName");
+
+		if (var != null) {
+			if (!(var instanceof String)) {
+				throw new IllegalArgumentException(
+						"Property FileName must be a String");
+			}
+
+			File indexFile = new File((String) var + ".idx");
+			File dataFile = new File((String) var + ".dat");
+
+			// check if files exist.
+			if (bOverwrite == false
+					&& (!indexFile.exists() || !dataFile.exists())) {
+				bOverwrite = true;
+			}
+
+			if (bOverwrite) {
+				if (indexFile.exists()) {
+					indexFile.delete();
+				}
+				if (dataFile.exists()) {
+					dataFile.delete();
+				}
+
+				boolean b = indexFile.createNewFile();
+				if (b == false) {
+					throw new IOException("Index file (" + (String) var
+							+ ".idx) cannot be opened.");
+				}
+
+				b = dataFile.createNewFile();
+				if (b == false) {
+					throw new IOException("Data file cannot be opened.");
+				}
+			}
+
+			m_indexFile = new RandomAccessFile(indexFile, "rw");
+			m_dataFile = new RandomAccessFile(dataFile, "rw");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when an if condition at line 133 evaluates to true, and an else condition at line 136 evaluates to true, the *DiskStorageManager* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *m\_dataFile* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).